### PR TITLE
Fixes cfg attributes in runtime macro

### PR DESCRIPTION
**Pull request by gupnik**
Fixes https://github.com/paritytech/polkadot-sdk/issues/6209

This PR adds the support for cfg attributes in the runtime macro.

### CODE REVIEW

File: substrate/frame/support/procedural/src/runtime/parse/mod.rs [hunk 1]
```diff
@@ -39,6 +39,7 @@ mod keyword {
 	custom_keyword!(pallet_index);
 	custom_keyword!(disable_call);
 	custom_keyword!(disable_unsigned);
+	custom_keyword!(disable_metadata);
```
Comment (bkchr) [inline substrate/frame/support/procedural/src/runtime/parse/mod.rs:42]: I don't think we need this. Without metadata you can not do anything.

**Review by bkchr — APPROVED** [review]
Besides the disable metadata stuff, it looks good.

File: substrate/frame/support/procedural/src/runtime/parse/mod.rs (hunk 1 above)
Comment (gupnik) [reply substrate/frame/support/procedural/src/runtime/parse/mod.rs:42]: > I don't think we need this. Without metadata you can not do anything.

Thanks - I  actually needed this for the test to ignore `frame_system`'s metadata. What do you suggest to be a good alternative?

Comment (gupnik) [reply substrate/frame/support/procedural/src/runtime/parse/mod.rs:42]: Thanks but this doesn't work. I think the type registries generate an index in the metadata which is incorrect if I just ignore the system pallet and try to match.

> Filter out the system metadata when asserting the metadata:
> 
> ```diff
> diff --git a/substrate/frame/support/test/tests/pallet.rs b/substrate/frame/support/test/tests/pallet.rs
> index 68efabce75e..7906a890593 100644
> --- a/substrate/frame/support/test/tests/pallet.rs
> +++ b/substrate/frame/support/test/tests/pallet.rs
> @@ -816,7 +816,6 @@ mod runtime {
>         pub struct Runtime;
>  
>         #[runtime::pallet_index(0)]
> -       #[runtime::disable_metadata]
>         pub type System = frame_system;
>  
>         #[runtime::pallet_index(1)]
> @@ -1954,7 +1953,7 @@ fn metadata() {
>                 _ => panic!("metadata has been bumped, test needs to be updated"),
>         };
>  
> -       pretty_assertions::assert_eq!(actual_metadata.pallets, expected_metadata.pallets);
> +       pretty_assertions::assert_eq!(actual_metadata.pallets[1..], expected_metadata.pallets);
>  }
>  
>  #[test]
> @@ -2569,8 +2568,8 @@ fn test_error_feature_parsing() {
>  fn pallet_metadata() {
>         use sp_metadata_ir::{DeprecationInfoIR, DeprecationStatusIR};
>         let pallets = Runtime::metadata_ir().pallets;
> -       let example = pallets[0].clone();
> -       let example2 = pallets[1].clone();
> +       let example = pallets[1].clone();
> +       let example2 = pallets[2].clone();
>         {
>                 // Example2 pallet is deprecated
>                 assert_eq!(
> ```

Comment (gui1117) [reply substrate/frame/support/procedural/src/runtime/parse/mod.rs:42]: Yes you can update the type registries with the resulting ones, we don't really assert against the type registry

Comment (gupnik) [reply substrate/frame/support/procedural/src/runtime/parse/mod.rs:42]: > Yes you can update the type registries with the resulting ones, we don't really assert against the type registry

Yeah, I don't think that it helps to change it once the metadata has been generated. However, I can see another way to remove this from metadata without adding `disable_metadata` attribute. Will add that and see if that works for you.

Comment (gupnik) [reply substrate/frame/support/procedural/src/runtime/parse/mod.rs]: Added. @gui1117 Could you please take a look now? Thanks

File: substrate/frame/support/procedural/src/runtime/parse/pallet.rs [hunk 2]
```diff
@@ -78,7 +78,18 @@ impl Pallet {
...
 
-		let cfg_pattern = vec![];
+		let cfg_pattern = item
+			.attrs
+			.iter()
+			.filter(|attr| attr.path().segments.first().map_or(false, |s| s.ident == "cfg"))
+			.map(|attr| {
+				attr.parse_args_with(|input: syn::parse::ParseStream| {
+					let input = input.parse::<proc_macro2::TokenStream>()?;
+					cfg_expr::Expression::parse(&input.to_string())
```
Comment (bkchr) [inline substrate/frame/support/procedural/src/runtime/parse/pallet.rs:88]:
```suggestion
					cfg_expr::Expression::parse(&input.to_string())
```

Should also work?

**Review by bkchr — APPROVED** [review]

**Review by gui1117 — APPROVED** [review]

**Pull request merged**